compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp

```diff
@@ -35,3 +37,290 @@ TEST(CloneNodeTest, clone_BatchMatMul)
+  ASSERT_EQ(4, shape.dim(2).value());
+  ASSERT_EQ(7, shape.dim(3).value());
+}
+
+TEST(ShapeRuleTest, bmm_not_broadcastable_1_NEG)
+{
+  luci::CircleInput input_x;
+  luci::CircleInput input_y;
+  luci::CircleBatchMatMul bmm;
+
+  input_x.shape({2, 4, 3});
+  input_x.shape_status(luci::ShapeStatus::VALID);
+
+  input_y.shape({3, 3, 7});
+  input_y.shape_status(luci::ShapeStatus::VALID);
+
+  bmm.x(&input_x);
+  bmm.y(&input_y);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  // (2, 4, 3) x (3, 3, 7)
+  //  ^           ^
+  // => error, batch dimension failed to broadcast
+  ASSERT_ANY_THROW(shape_inf_rule.infer(&bmm, shape));
+}
+
+TEST(ShapeRuleTest, bmm_not_broadcastable_2_NEG)
+{
+  luci::CircleInput input_x;
+  luci::CircleInput input_y;
+  luci::CircleBatchMatMul bmm;
+
+  input_x.shape({2, 4, 3});
+  input_x.shape_status(luci::ShapeStatus::VALID);
+
+  input_y.shape({1, 3, 3, 7});
+  input_y.shape_status(luci::ShapeStatus::VALID);
+
+  bmm.x(&input_x);
+  bmm.y(&input_y);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  // (2, 4, 3) x (1, 3, 3, 7)
+  //  ^           ^  ^
+  // => error, batch dimension failed to broadcast
+  ASSERT_ANY_THROW(shape_inf_rule.infer(&bmm, shape));
+}
+
+TEST(ShapeRuleTest, bmm_not_broadcastable_3_NEG)
+{
+  luci::CircleInput input_x;
+  luci::CircleInput input_y;
+  luci::CircleBatchMatMul bmm;
+
+  input_x.shape({2, 4, 3});
+  input_x.shape_status(luci::ShapeStatus::VALID);
+
+  input_y.shape({4, 3, 7});
+  input_y.shape_status(luci::ShapeStatus::VALID);
+
+  bmm.x(&input_x);
+  bmm.y(&input_y);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  // (2, 4, 3) x (4, 3, 7)
+  //  ^           ^
+  // => error, batch dimension failed to broadcast
+  ASSERT_ANY_THROW(shape_inf_rule.infer(&bmm, shape));
+}
+
+TEST(ShapeRuleTest, bmm_mismatch_dim_1_NEG)
+{
+  luci::CircleInput input_x;
+  luci::CircleInput input_y;
+  luci::CircleBatchMatMul bmm;
+
+  input_x.shape({1, 4, 4});
+  input_x.shape_status(luci::ShapeStatus::VALID);
+
+  input_y.shape({1, 3, 7});
+  input_y.shape_status(luci::ShapeStatus::VALID);
+
+  bmm.x(&input_x);
+  bmm.y(&input_y);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  // (1, 4, 4) x (1, 3, 7)
+  //        ^        ^
+  // => error, matmul hidden dim should be same
+  ASSERT_ANY_THROW(shape_inf_rule.infer(&bmm, shape));
+}
+
+TEST(ShapeRuleTest, bmm_mismatch_dim_2_NEG)
+{
+  luci::CircleInput input_x;
+  luci::CircleInput input_y;
+  luci::CircleBatchMatMul bmm;
+
+  input_x.shape({2, 40, 40});
+  input_x.shape_status(luci::ShapeStatus::VALID);
+
+  input_y.shape({1, 30, 70});
+  input_y.shape_status(luci::ShapeStatus::VALID);
+
+  bmm.x(&input_x);
+  bmm.y(&input_y);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  // (2, 40, 40) x (1, 30, 70)
+  //         ^         ^
+  // => error, matmul hidden dim should be same
+  ASSERT_ANY_THROW(shape_inf_rule.infer(&bmm, shape));
+}
```

Comment:
How about adding zero-dim cases but it is static. (actual 0-dim)
For example in python code,

```python
# Example 3: Edge case with zero-dimension batch
A = tf.random.normal([0, 2, 3])  # Shape: (0, 2, 3)
B = tf.random.normal([0, 3, 2])  # Shape: (0, 3, 2)
```
-> The shape of result should be (0, 2, 2)

and,

```python
# Example 4: Edge case with zero-dimension batch
A = tf.random.normal([2, 2, 3])  # Shape: (2, 2, 3)
B = tf.random.normal([0, 3, 2])  # Shape: (0, 3, 2)
```
-> The should generate an exception

Reply:
> How about adding zero-dim cases but it is static. (actual 0-dim) 

I guess if tensor has a 0 dimension, it means the tensor is empty. 
I thought that graph with empty tensor is invalid, and I didn't handle this case in shape inference. 

I'd better add an `assertion` or `exception` for this case. (also negative test case if possible). 
Thanks for catching.

Reply:
I updated the logic - https://github.com/Samsung/ONE/pull/13759/files#r1733812036, and test case - https://github.com/Samsung/ONE/pull/13759/files#r1733813949